### PR TITLE
[TTAHUB-484] Add program type scope for activity reports

### PR DESCRIPTION
**Pull request by thewatermethod**
## Description of change
Create two scopes on the backend that can be applied to activity reports, one to include activity reports on a given program type, one to exclude the same.

## How to test
This functionality is not implemented on the front end as of yet - read the code and the tests, make sure that the tests are written in a way as to actually test the scope, and that they pass

## Issue(s)

* https://ocio-jira.acf.hhs.gov/browse/TTAHUB-484

### Every PR

- [x] Meets issue criteria
- [x] JIRA ticket status updated
- [x] Code is meaningfully tested
- [n/a] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [n/a] API Documentation updated
- [n/a] Boundary diagram updated
- [n/a] Logical Data Model updated
- [n/a] [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions

### Production Deploy

- [ ] Staging smoke test completed

### After merge/deploy

- [ ] Update JIRA ticket status

### CODE REVIEW

File: src/scopes/activityReport/index.js [hunk 1]
```diff
@@ -13,6 +13,7 @@ import { withoutCalculatedStatus, withCalculatedStatus } from './calculatedStatu
 import { withProgramSpecialist, withoutProgramSpecialist } from './programSpecialist';
 import { withRole, withoutRole } from './role';
 import withRegion from './region';
+import { withoutProgramTypes, withProgramTypes } from './programType';
```
Comment (GarrettEHill) [inline src/scopes/activityReport/index.js:16]: Is there a reason that the import is "without" then "with", while use later is "with" then "without"?

File: src/scopes/activityReport/programType.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { Op } from 'sequelize';
+import { filterAssociation } from './utils';
+
+const programTypes = `
+    SELECT "ActivityRecipients"."activityReportId" FROM "ActivityRecipients"
+    INNER JOIN "Grants" ON "Grants".id = "ActivityRecipients"."grantId"
+    INNER JOIN "Programs" ON "Programs"."grantId" = "Grants"."id" 
+    WHERE "Programs"."programType"`;
```
Comment (GarrettEHill) [inline src/scopes/activityReport/programType.js:8]: What is the goal in the where, are you filtering out null values?

File: src/scopes/activityReport/index.test.js [hunk 3]
```diff
@@ -967,6 +975,164 @@ describe('filtersToScopes', () => {
...
+      const scope = filtersToScopes(filters);
+      const found = await ActivityReport.findAll({
+        where: { [Op.and]: [scope, { id: possibleIds }] },
+      });
+      expect(found.length).toBe(3);
+      expect(found.map((f) => f.id))
+        .toEqual(expect.arrayContaining([reportOne.id, reportTwo.id, reportThree.id]));
+    });
+
+    it('excludes program type', async () => {
```
Comment (jasalisbury) [inline src/scopes/activityReport/index.test.js:1124]: Can we add a test case that checks program type not being multiple values ie `programType.nin': ['EHS', 'HS']`?

File: src/scopes/activityReport/index.js (hunk 1 above)
Comment (thewatermethod) [reply src/scopes/activityReport/index.js:16]: Just me letting VS Code's auto-import decide

File: src/scopes/activityReport/index.test.js [hunk 4]
```diff
@@ -967,6 +975,164 @@ describe('filtersToScopes', () => {
     });
   });
 
+  describe('programType', () => {
```
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/scopes/activityReport/index.test.js:978]: Honestly forgot that the report creator helper existed. I will try it and see if using it cuts down on boilerplate

File: src/scopes/activityReport/utils.js [hunk 5]
```diff
@@ -23,15 +23,15 @@ export default function filterArray(column, searchTerms, exclude) {
   };
 }
 
-export function filterAssociation(baseQuery, searchTerms, exclude, comparator = '~*') {
+export function filterAssociation(baseQuery, searchTerms, exclude, comparator = '~*', operator = Op.and) {
   if (exclude) {
     return {
-      [Op.and]:
```
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/scopes/activityReport/utils.js:29]: I think you might be right, we always want ```Op.or```, actually... as long as the returned value from this function is always part of an ```Op.and``` already, which I think it is.

[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/scopes/activityReport/utils.js]: so after running some tests, I think we want [Op.or] for the include and [Op.and] for the exclude

File: src/scopes/activityReport/index.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (thewatermethod) [reply src/scopes/activityReport/index.test.js:1128]: Verdict: Cut down some boilerplate.

File: src/scopes/activityReport/programType.js (hunk 2 above)
Comment (thewatermethod) [reply src/scopes/activityReport/programType.js:9]: I did add some comments, although maybe a refactor would also be in order, not sure

**Review by GarrettEHill — APPROVED** [review]

**Review by jasalisbury — APPROVED** [review]
:shipit:

**Pull request merged**